### PR TITLE
8256368: Avoid repeated upcalls into Java to re-resolve MH/VH linkers/invokers

### PR DESCRIPTION
**Pull request by iwanowww**
Method linkage of `invokehandle` instructions involve an upcall into Java (`MethodHandleNatives::linkMethod`), but the result is not cached. Unless the upcall behaves idempotently (which is highly desirable, but not guaranteed), repeated invokehandle resolution attempts enter a vicious cycle in tiered mode: switching to a higher tier involves call site re-resolution in generated code, but re-resolution installs a fresh method which puts execution back into interpreter.

(Another prerequisite is no inlining through a `invokehandle` which doesn't normally happen in practice - relevant methods are marked w/ `@ForceInline` - except some testing modes, `-Xcomp` in particular.)

Proposed fix is to inspect `ConstantPoolCache` first. Previous resolution attempts from interpreter and C1 records their results there and it stabilises the execution.         

Testing:
 - failing test
 - tier1-4

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Testing

|     | Linux additional | Linux x64 | Linux x86 | Windows x64 | macOS x64 |
| --- | ----- | ----- | ----- | ----- | ----- |
| Build | ✔️ (8/8 passed) | ✔️ (2/2 passed) | ❌ (2/2 failed) | ✔️ (2/2 passed) | ❌ (2/2 failed) |
| Test (tier1) |    |  ❌ (1/9 failed) |    |  ✔️ (9/9 passed) |    | 

**Failed test tasks**
- [Linux x64 (hs/tier1 runtime)](https://github.com/iwanowww/jdk/runs/1459620381)
- [Linux x86 (build debug)](https://github.com/iwanowww/jdk/runs/1459213722)
- [Linux x86 (build release)](https://github.com/iwanowww/jdk/runs/1459213708)
- [macOS x64 (build debug)](https://github.com/iwanowww/jdk/runs/1459213936)
- [macOS x64 (build release)](https://github.com/iwanowww/jdk/runs/1459213916)

### Issue
 * [JDK-8256368](https://bugs.openjdk.java.net/browse/JDK-8256368): Avoid repeated upcalls into Java to re-resolve MH/VH linkers/invokers

### Reviewers
 * [Claes Redestad](https://openjdk.java.net/census#redestad) (@cl4es - **Reviewer**)
 * [Coleen Phillimore](https://openjdk.java.net/census#coleenp) (@coleenp - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1453/head:pull/1453`
`$ git checkout pull/1453`

### CODE REVIEW

**Review by cl4es — APPROVED** [review]
Nice!

File: src/hotspot/share/interpreter/linkResolver.cpp [hunk 1]
```diff
@@ -1706,15 +1699,35 @@ void LinkResolver::resolve_invokeinterface(CallInfo& result, Handle recv, const
   resolve_interface_call(result, recv, recvrKlass, link_info, true, CHECK);
 }
 
+bool LinkResolver::resolve_previously_linked_invokehandle(CallInfo& result, const LinkInfo& link_info, const constantPoolHandle& pool, int index, TRAPS) {
+  int cache_index = ConstantPool::decode_cpcache_index(index, true);
+  ConstantPoolCacheEntry* cpce = pool->cache()->entry_at(cache_index);
+  if (!cpce->is_f1_null()) {
```
Comment (coleenp) [inline src/hotspot/share/interpreter/linkResolver.cpp:1705]: If f1 is non-null any racing resolution is complete.  Can you double check if that's also the case for indy_resolution_failed?

**Review by coleenp — APPROVED** [review]
Looks great.  I had one question in the code.

File: src/hotspot/share/interpreter/linkResolver.cpp (hunk 1 above)
Comment (cl4es) [reply src/hotspot/share/interpreter/linkResolver.cpp:1705]: My reading is that a non-null f1 means any resolution races is complete and writes to indy_resolution_failed (part of flags) is done before writing f1, see comments in and logic in  [ConstantPoolCacheEntry::set_method_handle_common](https://github.com/openjdk/jdk/blob/7ed591cc9a1c46f7c0b10a4c4ad3c181333bf5b6/src/hotspot/share/oops/cpCache.cpp#L364). It thus seem consistent and correct to resolve successfully here, no?

**Pull request closed without merging**